game-core/src/main/java/games/strategy/triplea/delegate/AbstractEndTurnDelegate.java

```diff
@@ -64,10 +65,20 @@ private static boolean canPlayerCollectIncome(final PlayerID player, final GameD
    */
   public static IntegerMap<Resource> findEstimatedIncome(final PlayerID player, final GameData data) {
     final IntegerMap<Resource> resources = new IntegerMap<>();
-    final List<Territory> territories = data.getMap().getTerritoriesOwnedBy(player);
-    final int pusFromTerritories = getProduction(territories, data) * Properties.getPuMultiplier(data);
-    resources.add(new Resource(Constants.PUS, data), pusFromTerritories);
-    resources.add(EndTurnDelegate.getResourceProduction(territories, data));
+
+    // Only add territory resources if endTurn not endTurnNoPU
+    final Iterator<GameStep> it = data.getSequence().iterator();
+    while (it.hasNext()) {
+      final GameStep step = it.next();
```

Comment:
If we're not removing elements during iteration, we can probably collapse these three lines to a for-each loop:

```java
for (final GameStep step : data.getSequence()) {
```